examples/java/src/test/java/org/apache/beam/examples/complete/kafkatopubsub/KafkaToPubsubE2ETest.java

```diff
@@ -0,0 +1,119 @@
+import org.apache.kafka.clients.producer.KafkaProducer;
+import org.apache.kafka.clients.producer.ProducerConfig;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.common.serialization.StringSerializer;
+import org.joda.time.Duration;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.ClassRule;
+import org.junit.Rule;
+import org.junit.Test;
+import org.testcontainers.containers.KafkaContainer;
+import org.testcontainers.containers.PubSubEmulatorContainer;
+import org.testcontainers.utility.DockerImageName;
+
+/** E2E test for {@link KafkaToPubsub} pipeline. */
+public class KafkaToPubsubE2ETest {
+  private static final String PUBSUB_EMULATOR_IMAGE =
+      "gcr.io/google.com/cloudsdktool/cloud-sdk:316.0.0-emulators";
+  private static final String KAFKA_IMAGE_NAME = "confluentinc/cp-kafka:5.4.3";
+  private static final String PUBSUB_MESSAGE = "test pubsub message";
+  private static final String KAFKA_TOPIC_NAME = "messages-topic";
+  private static final String PROJECT_ID = "try-kafka-pubsub";
+  private static final PipelineOptions OPTIONS = TestPipeline.testingPipelineOptions();
+
+  @ClassRule public static final PubSubEmulatorContainer pubSubEmulatorContainer = new PubSubEmulatorContainer(DockerImageName.parse(PUBSUB_EMULATOR_IMAGE));
+  @ClassRule public static final KafkaContainer kafkaContainer = new KafkaContainer(DockerImageName.parse(KAFKA_IMAGE_NAME));
+  @Rule public final transient TestPipeline pipeline = TestPipeline.fromOptions(OPTIONS);
+  @Rule public final transient TestPubsub testPubsub = TestPubsub.fromOptions(OPTIONS);
+
+  @BeforeClass
+  public static void beforeClass() {
+    Credentials credentials = NoopCredentialFactory.fromOptions(OPTIONS).getCredential();
+    OPTIONS.as(DirectOptions.class).setBlockOnRun(false);
+    OPTIONS.as(GcpOptions.class).setGcpCredential(credentials);
+    OPTIONS.as(GcpOptions.class).setProject(PROJECT_ID);
+    OPTIONS.as(PubsubOptions.class).setPubsubRootUrl("http://" + pubSubEmulatorContainer.getEmulatorEndpoint());
+    OPTIONS.as(KafkaToPubsubOptions.class).setOutputFormat(FORMAT.PUBSUB);
+    OPTIONS.as(KafkaToPubsubOptions.class).setBootstrapServers(kafkaContainer.getBootstrapServers());
+    OPTIONS.as(KafkaToPubsubOptions.class).setInputTopics(KAFKA_TOPIC_NAME);
+    OPTIONS.as(KafkaToPubsubOptions.class).setKafkaConsumerConfig(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG + "=earliest");
+  }
+
+  @Before
+  public void setUp() {
+    OPTIONS.as(KafkaToPubsubOptions.class).setOutputTopic(testPubsub.topicPath().getPath());
+  }
+
+  @Test
+  public void testKafkaToPubsubE2E() throws IOException, InterruptedException {
+    PipelineResult job = KafkaToPubsub.run(pipeline, OPTIONS.as(KafkaToPubsubOptions.class));
+
+    sendKafkaMessage();
+    testPubsub.assertThatTopicEventuallyReceives(
+        hasProperty("payload", equalTo(PUBSUB_MESSAGE.getBytes(StandardCharsets.UTF_8)))
+    ).waitForUpTo(Duration.standardSeconds(25));
```

Comment:
25 seconds seem a bit unnatural. What do you think about replacing it with 30 seconds?

Reply:
I'll increase it to 1 minute